db/db_impl/db_impl.cc

```diff
@@ -1520,30 +1521,51 @@ Status DBImpl::GetImpl(const ReadOptions& read_options,
                         has_unpersisted_data_.load(std::memory_order_relaxed));
   bool done = false;
   if (!skip_memtable) {
-    if (sv->mem->Get(lkey, pinnable_val->GetSelf(), &s, &merge_context,
-                     &max_covering_tombstone_seq, read_options, callback,
-                     is_blob_index)) {
-      done = true;
-      pinnable_val->PinSelf();
-      RecordTick(stats_, MEMTABLE_HIT);
-    } else if ((s.ok() || s.IsMergeInProgress()) &&
-               sv->imm->Get(lkey, pinnable_val->GetSelf(), &s, &merge_context,
-                            &max_covering_tombstone_seq, read_options, callback,
-                            is_blob_index)) {
-      done = true;
-      pinnable_val->PinSelf();
-      RecordTick(stats_, MEMTABLE_HIT);
-    }
-    if (!done && !s.ok() && !s.IsMergeInProgress()) {
-      ReturnAndCleanupSuperVersion(cfd, sv);
-      return s;
+    if (get_val) {
+      if (sv->mem->Get(lkey, pinnable_val->GetSelf(), &s, &merge_context,
+                       &max_covering_tombstone_seq, read_options, callback,
+                       is_blob_index)) {
+        done = true;
+        pinnable_val->PinSelf();
+        RecordTick(stats_, MEMTABLE_HIT);
+      } else if ((s.ok() || s.IsMergeInProgress()) &&
+                 sv->imm->Get(lkey, pinnable_val->GetSelf(), &s, &merge_context,
+                              &max_covering_tombstone_seq, read_options,
+                              callback, is_blob_index)) {
+        done = true;
+        pinnable_val->PinSelf();
+        RecordTick(stats_, MEMTABLE_HIT);
+      }
+      if (!done && !s.ok() && !s.IsMergeInProgress()) {
+        ReturnAndCleanupSuperVersion(cfd, sv);
+        return s;
+      }
+    } else {
+      if (sv->mem->GetMergeOperands(lkey, pinnable_val, num_records, &s,
+                                    &merge_context, &max_covering_tombstone_seq,
+                                    read_options)) {
```

Comment:
The fact that GetImpl() sometimes does Get() and sometimes GetMergeOperands() is very confusing to me. Plus, pinnable_val is sometimes a pointer and sometimes an array. Do we have to do that, or can we create two separate functions?

Reply:
Creating 2 separate functions is causing code duplication hence I added new parameter called merge_operands in the existing function to hold merge operands if the call is requesting for merging operands and pinnable_val is reserved for holding value in KV.